eng/versioning.targets

```diff
@@ -20,6 +20,9 @@
     <AssemblyMetadata Include="PreferInbox">
       <Value>True</Value>
     </AssemblyMetadata>
+    <AssemblyMetadata Include="IsTrimmable">
```

Comment:
Are we sure we want to do this blindly? It feels like at least there should be a Condition to allow individual assemblies to turn it off.

Note that today we only analyze the assemblies in the shared fx / runtimepack for ILLink warnings. All OOB assemblies go unanalyzed. Once we have a story for analyzing libraries, we should hook `dotnet/runtime` up to it.

Reply:
I can add a condition. Are there any in particular that you think I should turn off to start with? (Would it maybe make sense to turn it off for OOB assemblies and opt them in individually when we analyze them?)

Reply:
Also, note that the feature as designed isn't tied to the warnings inherently - the attribute just says "trim me" - whether the assembly is linker-safe or not. That said, maybe it makes sense to opt out of trimming in cases where we don't trim a library during the runtime build because it uses too much reflection, like this:
https://github.com/dotnet/runtime/blob/75d5cc3df4dde3b37d75f289b5a90e358da838aa/src/libraries/System.Private.DataContractSerialization/src/System.Private.DataContractSerialization.csproj#L7-L8

Reply:
All libraries need to be trimmable in 6.0 but I'd probably exclude netfx build targets

Reply:
> All libraries need to be trimmable in 6.0

I agree we want to get there. But I think we should do it in a pragmatic way. Turning it on for all assemblies and letting people tell us what breaks doesn't seem like the best approach.

My thinking is that once we are done annotating the shared fx libraries, we start analyzing ILLink warnings for OOB libraries. When each library is "clean" of ILLink warnings, we mark it as "Trimmable".

For the initial change, I was assuming we would analyze System.IO.Pipelines, and assuming it is clean, we mark it as "IsTrimmable".

> like this: System.Private.DataContractSerialization.csproj

That one is part of the shared fx, so it gets trimmed today 🙃.

Reply:
We mark all non-OOB packages are trimmable today and in 5.0 and they have hundreds of linker warnings. Is there any reason why we need to handle OOB packages differently than non-OOB packages?

Reply:
AFAIK OOBs are currently not trimmed.

Reply:
I don't think the process we used in 5.0 was ideal. We didn't have the tools to do this the "right" way then. But now we do. 

We got a few bugs in 5.0 where someone would use a library no one had tested, and reported issues that it flat out didn't work once trimmed.

Reply:
> We got a few bugs in 5.0 where someone would use a library no one had tested, and reported issues that it flat out didn't work once trimmed.

Right and I think it's better to get that feedback during early previews where we can priorities better.

Relying on trimming warnings won't help because enabling trimming for libraryA can still break it, if libraryB is not trimming safe and assumes for example some attributes layout, internal members, etc

Reply:
Per discussion with @eerhardt, the plan is to opt all assemblies in as suggested by @marek-safar - and there's a new condition that we can use to opt out per project, should we need to.

By following the existing pattern for AssemblyMetadata like "Serviceable" and "PreferInbox", the attribute will be placed in src and ref assemblies, and also in assemblies targeting netstandard/netcoreapp/netfx. We need to include it for netstandard/netcoreapp because plenty of OOB assemblies build for those targets. Including the attributes in netfx assemblies shouldn't cause any issues since there are no common scenarios where an app targeting net6.0 references one of our netfx assemblies. I think it makes sense to be consistent with existing attributes on this, but @marek-safar if you prefer I could also disable it for netfx.

Reply:
> and ref assemblies

Are ref assemblies ever trimmed?

Reply:
> I think it makes sense to be consistent with existing attributes on this, but @marek-safar if you prefer I could also disable it for netfx.

I'm not against it, just wanted to be more cautious for netfx to not break bad code which could, for example, read only the last AssemblyMetadata but the chance is low.

Reply:
> Are ref assemblies ever trimmed?

No. But they are also never "serviced", and yet we are marking them with "Serviceable" assembly metadata. We should consistently apply these attributes across the assemblies.

Reply:
> No. But they are also never "serviced", and yet we are marking them with "Serviceable" assembly metadata. We should consistently apply these attributes across the assemblies.

FWIW starting .NET 5, we can now service the ref pack, we try to avoid it but we added support to be able to do it in case we need it, but I'm fine with adding IsTrimmable to them for consistency.